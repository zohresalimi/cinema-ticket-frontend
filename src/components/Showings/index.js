import React, { useContext } from "react";
import { Link } from "@reach/router";
import Moment from "react-moment";
import "moment-timezone";

import { useTranslation } from "react-i18next";
import AppContext from "../../store/context";

function getBookingText(showing) {
  if (!showing.capacity) return "sold out";
  if (new Date(showing.startTime) < new Date()) return "time passed";
  return "buy ticket";
}

function Showings({ category, currentMovie }) {
  const { state } = useContext(AppContext);
  const { cinemas, showings } = state;
  const { t } = useTranslation();

  return (
    <div data-testid="showings-wrapper">
      {cinemas.map((item) => {
        const showingsByCinema = showings.filter(
          (showing) => showing.room.cinema === item._id
        );
        return (
          <div key={item._id}>
            <h2 className="cinema-name">{item.name}</h2>
            <ul className="showings-list">
              {!!showingsByCinema.length &&
                showingsByCinema.map((el) => (
                  <li
                    key={el._id}
                    className={
                      getBookingText(el) !== "buy ticket" ? "unclickable" : ""
                    }
                  >
                    <Link
                      to={`../../../booking/${el._id}`}
                      state={{ category }}
                    >
                      <div className="showing-info">
                        <p className="time-slot">
                          <Moment
                            tz="Europe/Stockholm"
                            date={el.startTime}
                            format="hh:mm"
                          />
                        </p>
                        <div>
                          <p className="room-name">{t(el.room.name)}</p>
                          <p className="movie-subtitle">
                            {t("lang")}
                            <span>:</span>
                            {t(currentMovie.originalTitle)}
                          </p>
                          <p>
                            <Moment
                              tz="Europe/Stockholm"
                              date={el.startTime}
                              format="YYYY-MM-DD"
                            />
                          </p>
                        </div>
                      </div>
                      <div className="buy-ticket">
                        <p>{t(getBookingText(el))}</p>
                      </div>
                    </Link>
                  </li>
                ))}
            </ul>
          </div>
        );
      })}
    </div>
  );
}

export default Showings;
