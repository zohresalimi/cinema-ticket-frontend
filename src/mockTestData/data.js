import React from "react";
import { Router } from "@reach/router";
import AppContext from "../store/context";
import Theme from "../Styles/Theme";

export const getTestStore = () => {
  return {
    dispatch: () => {},
    state: {
      user: {
        name: "",
        email: "",
      },
      movies: {
        children: [
          {
            _id: "5fa71823adc463397a50c501",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "trolls 2: world tour",
            originalTitle: "swedish",
            premiere: "2020-10-03T15:18:52.722Z",
            rooms: ["5fa653d0a7f1201999dd1ac0"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        premiered: [
          {
            _id: "5fa71823adc463397a50c502",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "follow me",
            originalTitle: "swedish",
            premiere: "2020-10-05T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        upcoming: [
          {
            _id: "5fa71823adc463397a50c503",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "honest thief",
            originalTitle: "swedish",
            premiere: "2021-01-03T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        allMovies: {
          "5fa71823adc463397a50c501": "children",
          "5fa71823adc463397a50c502": "premiered",
          "5fa71823adc463397a50c503": "upcoming",
        },
      },
      cinemas: [
        {
          _id: "5fa64fe234fbbf9945f3a3ce",
          name: "filmstaden kista",
          purchaseEndTime: "21:30",
          purchaseStartTime: "8:00",
          rooms: ["5fa653c0643c06586bba7d78", "5fa653d0a7f1201999dd1ac0"],
        },
      ],
      ticket: {
        movie: {},
        status: false,
        quantity: 0,
        price: 0,
        seatNumbers: new Map(),
        user: {},
        showing: {},
      },
      showings: [
        {
          _id: "5fbe9bd31f58462b5c759b5c",
          capacity: 4,
          cinema: "5fa64fe234fbbf9945f3a3ce",
          endTime: null,
          movie: {
            _id: "5fa71823adc463397a50c502",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            name: "follow me",
            originalTitle: "swedish",
            premiere: "2020-10-05T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
          price: 100,
          room: { _id: "5fa653c0643c06586bba7d78" },

          seats: [
            [
              { seatNumber: 1, taken: "available" },
              { seatNumber: 1, taken: "available" },
            ],
            [
              { seatNumber: 1, taken: "available" },
              { seatNumber: 1, taken: "available" },
            ],
          ],
          startTime: "2020-11-13T16:00:00.000Z",
        },
      ],
    },
  };
};

export const WithProvider = (props) => {
  const { state, dispatch } = getTestStore();
  return (
    <Theme>
      <AppContext.Provider
        value={{ state: props.defaultValue || state, dispatch }}
      >
        <Router>{props.children}</Router>
      </AppContext.Provider>
    </Theme>
  );
};
