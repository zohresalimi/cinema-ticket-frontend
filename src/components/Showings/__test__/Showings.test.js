import React from "react";
import { render, act } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { getTestStore, WithProvider } from "../../../mockTestData/data";

import Showings from "../";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({
    t: (k) => k,
    i18n: {
      language: "sv",
      changeLanguage: () => new Promise(() => {}),
    },
  }),
}));

function renderWrapper(props) {
  let component;

  act(() => {
    component = render(
      <WithProvider>
        <Showings {...props} path="/" />
      </WithProvider>
    );
  });

  return component;
}

describe("Showings Component Testing", () => {
  const cinemas = getTestStore().cinemas;
  const showings = getTestStore().showings;
  const currentMovie = getTestStore().movies.children[0];
  test("take snapshot when cinema & showings do not match", async () => {
    const { container } = await renderWrapper({
      cinemas,
      showings,
      category: "children",
      currentMovie,
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test("take snapshot when cinema & showings do match", async () => {
    showings[0].room = { ...showings[0].room, cinema: cinemas[0]._id };
    const { container } = await renderWrapper({
      cinemas,
      showings,
      category: "children",
      currentMovie,
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
