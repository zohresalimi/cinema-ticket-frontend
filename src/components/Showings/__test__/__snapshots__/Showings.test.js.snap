// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Showings Component Testing take snapshot when cinema & showings do match 1`] = `
<div
  style="outline: none;"
  tabindex="-1"
>
  <div
    data-testid="showings-wrapper"
  >
    <div>
      <h2
        class="cinema-name"
      >
        filmstaden kista
      </h2>
      <ul
        class="showings-list"
      >
        <li>
          <a
            href="/booking/5fbe9bd31f58462b5c759b5c"
          >
            <div
              class="showing-info"
            >
              <p
                class="time-slot"
              >
                <time
                  datetime="1605283200000"
                >
                  05:00
                </time>
              </p>
              <p
                class="room-name"
              />
              <p
                class="movie-subtitle"
              >
                lang
                <span>
                  :
                </span>
                swedish
              </p>
            </div>
            <div
              class="buy-ticket"
            >
              <p>
                buy ticket
              </p>
            </div>
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Showings Component Testing take snapshot when cinema & showings do not match 1`] = `
<div
  style="outline: none;"
  tabindex="-1"
>
  <div
    data-testid="showings-wrapper"
  >
    <div>
      <h2
        class="cinema-name"
      >
        filmstaden kista
      </h2>
      <ul
        class="showings-list"
      />
    </div>
  </div>
</div>
`;
